### ISSUE TITLE
MultipleObjectsReturned: get() returned more than one Embed

### ISSUE THREAD

**Issue by lee-kagiso** [wagtail/wagtail#1745]
We are getting this quite a lot.

We have  a Sentry report over here: https://beta.getsentry.com/share/group/34313638322e3833333433303131/

```
MultipleObjectsReturned: get() returned more than one Embed -- it returned 2!
  File "django/core/handlers/base.py", line 137, in get_response
    response = response.render()
  File "django/template/response.py", line 103, in render
    self.content = self.rendered_content
  File "django/template/response.py", line 80, in rendered_content
    content = template.render(context)
  File "django/template/base.py", line 148, in render
    return self._render(context)
  File "django/template/base.py", line 142, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 844, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 858, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 126, in render
    return compiled_parent._render(context)
  File "django/template/base.py", line 142, in _render
    return self.nodelist.render(context)
  File "django/template/base.py", line 844, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 858, in render_node
    return node.render(context)
  File "django/template/loader_tags.py", line 65, in render
    result = block.nodelist.render(context)
  File "django/template/base.py", line 844, in render
    bit = self.render_node(node, context)
  File "django/template/base.py", line 858, in render_node
    return node.render(context)
  File "django/template/base.py", line 904, in render
    return render_value_in_context(output, context)
  File "django/template/base.py", line 881, in render_value_in_context
    value = force_text(value)
  File "django/utils/encoding.py", line 83, in force_text
    s = six.text_type(s)
  File "wagtail/wagtailcore/blocks/stream_block.py", line 311, in __str__
    return self.stream_block.render(self)
  File "wagtail/wagtailcore/blocks/base.py", line 216, in render
    return self.render_basic(value)
  File "wagtail/wagtailcore/blocks/stream_block.py", line 201, in render_basic
    [(force_text(child), child.block_type) for child in value]
  File "wagtail/wagtailcore/blocks/stream_block.py", line 201, in <listcomp>
    [(force_text(child), child.block_type) for child in value]
  File "django/utils/encoding.py", line 83, in force_text
    s = six.text_type(s)
  File "wagtail/wagtailcore/blocks/stream_block.py", line 251, in __str__
    return self.block.render(self.value)
  File "wagtail/wagtailcore/blocks/base.py", line 216, in render
    return self.render_basic(value)
  File "wagtail/wagtailcore/blocks/base.py", line 223, in render_basic
    return force_text(value)
  File "django/utils/encoding.py", line 83, in force_text
    s = six.text_type(s)
  File "wagtail/wagtailembeds/blocks.py", line 21, in __str__
    return embed_to_frontend_html(self.url)
  File "wagtail/wagtailembeds/format.py", line 10, in embed_to_frontend_html
    embed = embeds.get_embed(url)
  File "wagtail/wagtailembeds/embeds.py", line 128, in get_embed
    return Embed.objects.get(url=url, max_width=max_width)
  File "django/db/models/manager.py", line 92, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "django/db/models/query.py", line 361, in get
    num if num <= MAX_GET_RESULTS else 'more than %s' % MAX_GET_RESULTS
```

**Comment by kaedroho** [comment]
Thanks! It looks like we either need to add a unique constraint, or make the ``get_embed`` function handle the error better.

The ``Embed`` model is just a cache, so it should be safe to delete the duplicate in the database. Any missing ``Embed`` records will be regenerated.

**Comment by gasman** [comment]
@lee-kagiso I take it you're using SQLite as your database backend (based on the lack of postgresql / mysql client library under Packages in the Sentry report)?

We have a `unique_together` constraint on the Embed model which *should* prevent these duplicates from occurring at the database level: https://github.com/torchbox/wagtail/blob/8bf9851e01e8a5128cfc9d235b165d23954fdb41/wagtail/wagtailembeds/models.py#L41. However, according to http://stackoverflow.com/questions/8844283/django-unique-together-not-preventing-duplicates these are not respected by SQLite. If so, I don't think there's much we can do to avoid this on SQLite.

**Comment by lee-kagiso** [comment]
We are using Postgres...

**Comment by gasman** [comment]
Hmm... what do you get if you enter `\d wagtailembeds_embed` at the `psql` command line? (I'm particularly interested to know if there's a `UNIQUE CONSTRAINT, btree (url, max_width)` line at the bottom.)

**Comment by kaedroho** [comment]
Is ``max_width`` set to``None`` in the duplicates? I'm not sure if the ``unique_together`` constraint is enforced if one of the fields doesn't have a value.

**Comment by lee-kagiso** [comment]
I get this:

```
kagisocms=# \d wagtailembeds_embed
                                      Table "public.wagtailembeds_embed"
+---------------+--------------------------+------------------------------------------------------------------+
|    Column     |           Type           |                            Modifiers                             |
+---------------+--------------------------+------------------------------------------------------------------+
| id            | integer                  | not null default nextval('wagtailembeds_embed_id_seq'::regclass) |
| url           | character varying(200)   | not null                                                         |
| max_width     | smallint                 |                                                                  |
| type          | character varying(10)    | not null                                                         |
| html          | text                     | not null                                                         |
| title         | text                     | not null                                                         |
| author_name   | text                     | not null                                                         |
| provider_name | text                     | not null                                                         |
| thumbnail_url | character varying(200)   |                                                                  |
| width         | integer                  |                                                                  |
| height        | integer                  |                                                                  |
| last_updated  | timestamp with time zone | not null                                                         |
+---------------+--------------------------+------------------------------------------------------------------+
Indexes:
    "wagtailembeds_embed_pkey" PRIMARY KEY, btree (id)
    "wagtailembeds_embed_url_12c617386ccf42c1_uniq" UNIQUE CONSTRAINT, btree (url, max_width)
```

**Comment by gasman** [comment]
@lee-kagiso Thanks!

@kaedroho Argh, yes, you're right - have just confirmed that Postgres will let me insert multiple records for the same URL with a null `max_width`. As far as I'm aware, the race condition protection in `get_or_create` relies on the database enforcing this - very annoying. I guess the most sensible solution is to use something like -1 as a 'magic number' in place of null. (I suggest -1 rather than 0, as 0 might conceivably be a valid parameter - if you want an invisible audio embed, for example.)

**Comment by gasman** [comment]
I believe this is now fixed as of #6635 and #6815 - #6635 replaced the old unique_together constraint (with its unexpected treatment of null fields) with a unique constraint on the `hash` field (where that quirk doesn't arise), and #6815 cleans up the duplicate records that were presumably created as a result of this bug.
